### PR TITLE
Set the default for a realm without depending on the representation

### LINKED ISSUE #14583: Provide partial import of realms for the map storage, ideally without needing a representation
### Description

When implementing #13609, it became apparent that `RealmManager.importRealm()` contains a lot of logic that is applied to a partial realm that contains at minimum only the enabled flag and the name of the realm. 

While a full import for the map storage wouldn't need that, a partial import for the map store would need that. According to @hmlnarik a partial realm definition for the file map store should be possible.

Therefore the goal of this issue is to:

* Provide the same logic as in `RealmManager.importRealm()` for the map storage by interacting on the model, and not on a representation.
* As necessary for the above, move over the necessary bits of `RealmManager.importRealm()` and `RealmManager.createRealm()`
* Consolidate `RealmManager.createRealm()` and `RealmManager.importRealm()` as much as possible, eventually having only a `importPartialRealm()` method

Making this work without a representation could enable the import of a partial file store to add the necessary defaults without the need to create a RealmRepresentation first.

### Motivation

Enable partial import for Map filestore.

### Details

When filling the physical store, it might not be possible for some store to store only partial information, some fields might be required. For testing, the CHM store might have the least requirements. 

It might be possible/necessary to first import to CHM, then augment the necessary bits, and finally to copy over the contents to the real store (using DeepCloner as necessary).

### PR DESCRIPTION
**Pull request by ahus1**
This enabled running defaults also for partial imports on the file store.

Closes #14583

### CODE REVIEW

**Comment by ahus1** [comment]
@hmlnarik - I made some progress here and I'd like to discuss to see if it goes into the right direction: 

* MapExportImportManager takes care of getting the data from the representation into the store
* A new event is fired, and all providers can then add their defaults to the realm
* The RealmManagerProviderFactory then acts on the realm to do the necessary bits. 

I went for an event to be able to call things from the services layer. It started with constants like `(OIDCLoginProtocol.LOGIN_PROTOCOL` and continued with `DefaultClientRegistrationPolicies.addDefaultPolicies(realm)`, so keeping everything in the map module wasn't a choice here. 

For now, this works fine with the JPA map store, and testing it manually I was able to import a full export, partial import and a new realm from the UI. 

I found that some defaults are already set when importing some flows, for example `DefaultAuthenticationFlows.migrateFlows()`. I assume we'll find those once there is a file store import to test with.

File: model/map/src/main/java/org/keycloak/models/map/datastore/MapExportImportManager.java [hunk 1]
```diff
@@ -441,7 +448,107 @@ public RealmModel importRealm(InputStream requestBody) {
...
+        try {
+            session.getContext().setRealm(realm);
+            ReservedCharValidator.validate(rep.getRealm());
+            realm.setName(rep.getRealm());
+
+            // setup defaults that will be overwritten by the import if present in the import
+            setupRealmDefaults(realm);
+
+            /// THIS IS THE REAL IMPORT ///
+            RepresentationToModel.importRealm(session, rep, realm, false);
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/datastore/MapExportImportManager.java:470]: This is actually calling 
```java
session.getProvider(DatastoreProvider.class).getExportImportManager().importRealm(rep, newRealm, skipUserDependent);
```
Should this line be thus replaced with the following?
```suggestion
            importRealm(rep, newRealm, skipUserDependent);
```

File: model/map/src/main/java/org/keycloak/models/map/datastore/MapExportImportManager.java [hunk 2]
```diff
@@ -441,7 +448,107 @@ public RealmModel importRealm(InputStream requestBody) {
...
+            setupClientServiceAccountsAndAuthorizationOnImport(rep, realm,false);
+            // why does this happen here?
+            session.clientPolicy().updateRealmModelFromRepresentation(realm, rep);
```
Comment (hmlnarik) [inline model/map/src/main/java/org/keycloak/models/map/datastore/MapExportImportManager.java:473]: I have no clue, but @mposolda might have.

**Review by hmlnarik — COMMENTED** [review]
Thanks @ahus1! I believe this PR is a good first step in the right direction

**Comment by ahus1** [comment]
@hmlnarik - I started out with good intentions, still with the existing JSON realms in the test suite this seems to be a futile endeavor: 

* The existing realm import needs some pre-sets (for example client scopes, but only if they are not part of the JSON), otherwise the import will skip the references to those in the import (and log a warning)
* After importing the representation, there's a step that might add the service accounts for a client; those classes are not available from the map module (and probably shouldn't be)
* As I final step I added setting the missing defaults by only looking at the realm, and not at the representation. It will behave differently as before: Before, it looked at the representation, and if the list was null it would add the defaults. Now, it looks if the list is empty in the model, and then creates the default. So there is no more the possibility to pass an empty list in the JSON to not create the default values. 

Looking at this, it doesn't seems to be a good idea to 
* change any of the existing behavior (it is undocumented and not covered by tests to avoid regressions), and 
* use this behavior for a future file store.

I wonder if the file store could pass something similar to a schema version that would define which defaults it would expect, and how one could choose between an empty and a default list of for example clients or client scopes. And then we could eventually have something like a migration of the then imported realm that would add missing bits and pieces for newer versions of Keycloak.

Could we start with writing the docs / the behavior as seen by the user to find something workable? For example, should a client with a service account be importable without a service account? Or with client scopes, when the client scopes are not part of the realm definition passed in the InputStream?

**Comment by ahus1** [comment]
Made some progress; now a lot of tests pass after importing a partial realm into a realm pre-initalized with defaults. Most of the time I needed to remove the exiting entities by name (except for the RSA key, that was a bit different). 

Will look at fixing some more tests on Monday, but not too much as I expect we're throwing it away, anyway, as we don't want to change the behavior of importing legacy JSONs (to be discussed on Monday).

I'm looking forward to discuss how I could do that with a CHM.

**Comment by ahus1** [comment]
In commit 4d84a53f2442fafa819880c7f08791e2de614679 I managed to import the realm representation to a CHM store and then copy it to the JPA map store. 

In order to force the creation of a CHM, I needed to wrap the session and the factory.

Tested manually with Untertow. The tests on Quarkus will probably all fail as the CHM factory will not be instantiated AFAIK when used with JPA Map provider. 

Looking forward to discuss the next steps next week.

**Comment by ahus1** [comment]
Hi @hmlnarik - the commit 862b65f7fcfdf8a80d7b1079e375fd4684e65bb0 contained the final changed discussed earlier today. It's now all squashed and ready for the final review and merge. Thanks!

**Review by hmlnarik — APPROVED** [review]
Thank you for the PR!

**Pull request merged**